Empire daemon

mwildehahn opened the pull request:
Still testing this, but this upgrades ECS and docker and removes empire to prep running as a daemon.

mwildehahn committed:
Run empire as a daemon on ECS

This removes empire from being run via the AMI on the controller. We now
use the AMI to setup instances that will run within minion and
controller clusters.

mwildehahn committed:
Bump ECS agent to 1.10.0

mwildehahn committed:
Update docker to 1.11.2

mwildehahn committed:
Remove old docker.conf file and setup new docker.default

mwildehahn committed:
Move docker.default to a template dir

ejholmes commented on ansible/roles/empire/templates/etc/empire/env/empire.env:
```diff
@@ -1,27 +0,0 @@
-EMPIRE_PORT=8080
```
I think we also need to update the [readme](https://github.com/remind101/empire_ami/blob/master/README.md)

ejholmes commented:
At this point, the AMI is almost identical to the official Amazon ECS AMI (granted, that one isn't ubuntu). The only real difference here is that we include logspout.

I wonder if it would be better for us to just open source an ansible role for installing logspout. Thoughts?

mwildehahn commented on ansible/roles/docker/templates/etc/default/docker.default:
```diff
@@ -4,8 +4,11 @@
+# dnsmasq running on the host. We want to do this to provide access to hosts
+# we write to /etc/hosts, in particular the dockerhost entry, where statsd will
+# be accessible from.
+DOCKER_OPTS="--dns {{ansible_default_ipv4.address}}"
```
this was causing:

```
docker logs -f 33fb1d9a9b94
2016-06-18T01:21:37Z [INFO] Starting Agent: Amazon ECS Agent - v1.10.0 (9a1b0f6)
2016-06-18T01:21:37Z [INFO] Loading configuration
2016-06-18T01:21:37Z [WARN] Invalid value for task cleanup duration, will be overridden to 3h0m0s, parsed value 0, minimum threshold 1m0s
2016-06-18T01:21:37Z [INFO] Checkpointing is enabled. Attempting to load state
2016-06-18T01:21:37Z [INFO] Loading state! module="statemanager"
2016-06-18T01:21:37Z [INFO] Detected Docker versions [1.17 1.18 1.19 1.20 1.21]
2016-06-18T01:21:37Z [INFO] Registering Instance with ECS
2016-06-18T01:21:58Z [ERROR] Could not register module="api client" err="RequestError: send request failed
caused by: Post https://ecs.us-west-2.amazonaws.com/: net/http: request canceled while waiting for connection"
2016-06-18T01:21:58Z [ERROR] Error registering: RequestError: send request failed
caused by: Post https://ecs.us-west-2.amazonaws.com/: net/http: request canceled while waiting for connection
```

removing this file completely fixed the issue

mwildehahn committed:
Disable --dns entry

phobologic commented:
@ejholmes So are you suggesting that we drop default logging support in the open source AMI, and make people build their own AMI with logspout, etc if they want that feature?

ejholmes commented:
Pretty much. Just not sure how much value this AMI actually provides now that almost everything is stripped out.

phobologic commented:
I think getting rid of logspout is kind of a big deal - it means that a really cool feature of empire won't exist for most users, and people aren't likely to want to build their own AMI. The good news is that the AMI should be much simpler after getting Empire off it, so I don't think maintaining it will be as big a deal

phobologic commented:
Also, if we eventually go to the logstash based logger, we'll need to install dnsmasq as well, so we'll need some level of customization anyway.  Not sure using the ECS image is going to work well for us going forward.

ejholmes commented:
Yeah that's true. Although, I was thinking we could just give some example cloudconfig to run logspout for `emp logs` support.

Logging is such a specific thing, it's hard to keep it generic enough to warrant maintaining it.

phobologic commented:
Yeah, I would agree if it weren't a feature of Empire itself, and one that I think a lot of users would be bummed to lose.

The problem with using cloudconfig is that it's done on boot, so you're just asking for it to fail every once in a while and end up with hosts not behaving how you like.

ejholmes commented:
Yeah true. If we stick with this, then we should aim to replace the Amazon ECS AMI in https://github.com/remind101/empire/blob/master/docs/cloudformation.json#L59-L63

phobologic commented:
Yeah, fair - having them be different does sort of suck, and it'd be cool to get logging working in the demo env.

mwildehahn committed:
Install ansible 2.1

mwildehahn committed:
Remove tailing whitespace

mwildehahn committed:
enable_streaming_logs isn't used by the controller

mwildehahn committed:
Update the README

mwildehahn committed:
Update CHANGELOG

mwildehahn committed:
Use --restart=always instead of respawn for the ecs-agent upstart job

mwildehahn committed:
Revert "Use --restart=always instead of respawn for the ecs-agent upstart job"

This reverts commit a9ecd48fd0d466975847dcd733dd2a9e9ba489a0.

mwildehahn committed:
Add a fork of docker.conf with a patch to handle docker.sock being a directory

There are cases where /var/run/docker.sock can get mounted as a
directory due to a race condition when restarting the docker daemon and
docker containers that mount /var/run/docker.sock. If the container
attempts to mount /var/run/docker.sock and it doesn't exist, it will
default it to a directory (similar to this thread:
https://bugzilla.redhat.com/show_bug.cgi?id=1289851). We can't use
`--restart=always` since it doesn't allow you to provide a name for the
container, which means we wouldn't be able to use upstart to stop the
container by name.

phobologic commented:
@mhahn, can you take a look at this? 

https://github.com/remind101/empire_ami/pull/20/files#diff-c478b6b4b86b0ea4f627fd66bf6eef25R2

Thanks @gswallow

mwildehahn committed:
Use /var/lib

/var/run gets blown away with every instance reboot. This results in the
ecs agent registering itself as a new instance everytime the minion is
restarted

mwildehahn committed:
Remove `notify` directives that break the upstart dependency chain

This resolves the issue that led to us forking the `docker.conf` file
(ie. /var/run/docker.sock being mounted as a directory due to a race
condition)

mwildehahn committed:
Use upstart dependency chain

mwildehahn commented:
@phobologic yea done

working as expected now

mwildehahn commented:
actually, @gswallow, i'm still seeing the same issue when i reboot using `/var/lib`. can you verify that it's not happening for you guys even when using `/var/lib`?

gswallow commented:
Which version of the agent are you using?  There's a known issue that was
fixed in ECS agent version 1.8.1.

![reboot](https://cloud.githubusercontent.com/assets/4225960/16341902/63dc1d92-39fd-11e6-9024-1fdf6b73af2f.png)

mwildehahn commented:
I'm running: 1.10.0

phobologic commented on ansible/roles/docker/templates/etc/default/docker.default:
```diff
@@ -4,8 +4,11 @@
 
 # Use DOCKER_OPTS to modify the daemon startup options.
-#DOCKER_OPTS="--dns 8.8.8.8 --dns 8.8.4.4"
-DOCKER_OPTS="--storage-driver=devicemapper"
```
Removing this moves us back to aufs right?  I think that's the right call.

mwildehahn replied:
Yea

```
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 26
 Dirperm1 Supported: false
```

I'm not familiar with all the different storage drivers, but after reading through the descriptions, it seems like `aufs` is the best fit for empire

phobologic replied:
👍

gswallow replied:
We can also make the switch to aufs.  How many docker containers do you find yourselves running on each minion, if I may ask?

phobologic replied:
We run around 10-15 - I think we run anywhere from 450-500 docker containers in a cluster of 35 minions (c4.2xlarge). We'd probably run more if our main API/mega-ruby-app didn't need 6 gigs of RAM to run :)

phobologic commented on ansible/roles/docker/templates/etc/default/docker.default:
```diff
@@ -4,8 +4,11 @@
-DOCKER_OPTS="--storage-driver=devicemapper"
+# We are adding the docker host IP as the dns server. This will be serviced by
+# dnsmasq running on the host. We want to do this to provide access to hosts
+# we write to /etc/hosts, in particular the dockerhost entry, where statsd will
```
Leftover comment/opts.

phobologic commented:
👍

mwildehahn committed:
Remove docker.default file

mwildehahn committed:
Use /var/lib/ instead of /var/run

/var/run gets blown away when rebooting the ec2 instance which can cause
the ecs agent to re-register itself for the same instance. /var/lib is
persisted between restarts.

mwildehahn commented:
@gswallow 🙈 i hadn't swapped the ecs_agent docker image from using `/var/run`. thanks!

mwildehahn commented:
verified that reboot doesn't reregister the instance
